gcc/rust/util/rust-punycode.cc

```diff
@@ -0,0 +1,180 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+// This file provides functions for punycode conversion
+// See https://datatracker.ietf.org/doc/html/rfc3492
+
+#include "rust-system.h"
+#include "rust-unicode.h"
+#include "optional.h"
+#include "selftest.h"
+
+namespace Rust {
+
+// https://tools.ietf.org/html/rfc3492#section-4.
+constexpr uint32_t BASE = 36;
+constexpr uint32_t TMIN = 1;
+constexpr uint32_t TMAX = 26;
+constexpr uint32_t SKEW = 38;
+constexpr uint32_t DAMP = 700;
+constexpr uint32_t INITIAL_BIAS = 72;
+constexpr uint32_t INITIAL_N = 128;
+constexpr char DELIMITER = '-';
+
+constexpr uint32_t MAX_ASCII_CODEPOINT = 0x7F;
+
+std::string
+extract_basic_string (const std::vector<Codepoint> &src)
+{
+  std::string basic_string;
+  for (auto c : src)
+    {
+      if (c.value <= MAX_ASCII_CODEPOINT)
+	basic_string += c.as_string ();
+    }
+  return basic_string;
+}
+
+uint32_t
+adapt_bias (uint32_t delta, const uint32_t n_points, const bool is_first)
+{
+  delta /= is_first ? DAMP : 2;
+  delta += delta / n_points;
+  uint32_t k = 0;
+
+  while (delta > (BASE - TMIN) * TMAX / 2)
+    {
+      delta /= BASE - TMIN;
+      k += BASE;
+    }
+  return k + (BASE - TMIN + 1) * delta / (delta + SKEW);
+}
+
+uint32_t
+clamped_sub (const uint32_t min, const uint32_t lhs, const uint32_t rhs,
+	     const uint32_t max)
+{
+  if (min + rhs >= lhs)
+    return min;
+  else if (max + rhs <= lhs)
+    return max;
+  else
+    return lhs - rhs;
+}
+
+uint32_t
+min_gt_or_eq (const std::vector<Codepoint> &l, const uint32_t threshold)
+{
+  uint32_t min = UINT32_MAX;
+  for (auto c : l)
+    if (c.value >= threshold && c.value < min)
+      min = c.value;
+  return min;
+}
+
+char
+encode_digit (const uint32_t d)
+{
+  return d + 22 + (d < 26 ? 75 : 0);
```

Comment:
I'd like some comments here to describe the behavior of this function. Newcomers should be able to understand why and where do these values come from.

Reply:
This function comes from the reference implementation in the RFC
https://datatracker.ietf.org/doc/html/rfc3492

```
/* encode_digit(d,flag) returns the basic code point whose value      */
/* (when used for representing integers) is d, which needs to be in   */
/* the range 0 to base-1.  The lowercase form is used unless flag is  */
/* nonzero, in which case the uppercase form is used.  The behavior   */
/* is undefined if flag is nonzero and digit d has no uppercase form. */

static char encode_digit(punycode_uint d, int flag)
{
  return d + 22 + 75 * (d < 26) - ((flag != 0) << 5);
  /*  0..25 map to ASCII a..z or A..Z */
  /* 26..35 map to ASCII 0..9         */
}
```

These value comes from a bit complicated calculation on parameters specific to Punycode.
I do not completely understand it but I think it is fine as is.

Reply:
Oh ok. Given you've already provided the link at the top of the file I don't think there's much we can do.